src/System.Text.Json/tests/Serialization/Value.ReadTests.cs

```diff
@@ -380,12 +380,62 @@ private static void AssertFloatingPointBehavior<T>(T netfxExpectedValue, T netco
         public static void LongInputString(int length)
         {
             // Verify boundary conditions in Deserialize() that inspect the size to determine allocation strategy.
-            string repeated = new string('x', length - 2);
-            string json = $"\"{repeated}\"";
-            Assert.Equal(length, json.Length);
+            DeserializeLongJsonString(length);
+        }
+
+        private const int MaxInt = int.MaxValue / MaxExpansionFactorWhileTranscoding;
+        private const int MaximumPossibleStringLength = int.MaxValue / 2 - 32;
+
+        // NOTE: VeryLongInputString test is constrained to run on Windows and MacOSX because it causes
+        //       problems on Linux due to the way deferred memory allocation works. On Linux, the allocation can
+        //       succeed even if there is not enough memory but then the test may get killed by the OOM killer at the
+        //       time the memory is accessed which triggers the full memory allocation.
+        [ConditionalTheory(nameof(IsX64))]
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.OSX)]
+        [InlineData(MaxInt - 3)]
+        [InlineData(MaxInt - 2)]
+        [InlineData(MaxInt - 1)]
+        [InlineData(MaxInt)]
+        [InlineData(MaxInt + 1)]
+        [InlineData(MaxInt + 2)]
+        [InlineData(MaxInt + 3)]
+        [InlineData(MaximumPossibleStringLength - 3)]
+        [InlineData(MaximumPossibleStringLength - 2)]
+        [InlineData(MaximumPossibleStringLength - 1)]
+        [InlineData(MaximumPossibleStringLength)]
+        [OuterLoop]
+        public static void VeryLongInputString(int length)
+        {
+            // Verify that deserializer does not do any multiplication or addition on the string length
+            DeserializeLongJsonString(length);
+        }
+
+        private static void DeserializeLongJsonString(int stringLength)
+        {
+            string json, str;
+
+#if BUILDING_INBOX_LIBRARY
+            char fillChar = 'x';
+            json = string.Create(stringLength, fillChar, (chars, fillChar) =>
+            {
+                chars.Fill(fillChar);
+                chars[0] = '"';
+                chars[chars.Length - 1] = '"';
+            });
 
-            string str = JsonSerializer.Deserialize<string>(json);
+            Assert.Equal(stringLength, json.Length);
+
+            str = JsonSerializer.Deserialize<string>(json);
+            Assert.True(json.AsSpan(1, json.Length - 2).SequenceEqual(str));
```

Comment:
These lines can be extracted out of the if/def and used for both platforms to avoid duplication since it doesn't contain .NET Core 3.0 specific code.

The string equals within `Assert.Equal(repeated, str);` is effectively doing a span.SequenceEqual under the covers anyway.
https://github.com/dotnet/corefx/blob/48363ac826ccf66fbe31a5dcb1dc2aab9a7dd768/src/Common/src/CoreLib/System/String.Comparison.cs#L28-L39